rpc/common/TransactionRequestBatcher.java

```diff
@@ -0,0 +1,170 @@
+    }
+
+    public Executor nextExecutor() {
+        return executors.get(executorIndex.getAndUpdate(i -> {
+            i++; if (i % executors.size() == 0) i = 0; return i;
+        }));
+    }
+
+    @Override
+    public void close() {
+        try {
+            accessLock.writeLock().lock();
+            if (isOpen) {
+                isOpen = false;
+                executors.forEach(Executor::clear);
+            }
+        } finally {
+            accessLock.writeLock().unlock();
+        }
+    }
+
+    public class Executor {
+
+        private final ConcurrentSet<Dispatcher> dispatchers;
+        private final AtomicBoolean isRunning;
+        private final Semaphore permissionToRun;
+
+        private Executor(ThreadFactory threadFactory) {
+            dispatchers = new ConcurrentSet<>();
+            isRunning = new AtomicBoolean(false);
+            permissionToRun = new Semaphore(0);
+            threadFactory.newThread(this::run).start();
+        }
+
+        private void mayStartRunning() {
+            if (isRunning.compareAndSet(false, true)) permissionToRun.release();
+        }
+
+        private void run() {
+            while (isOpen) {
+                try {
+                    permissionToRun.acquire();
+                    while (true) {
+                        Thread.sleep(BATCH_WINDOW_MILLIS);
+                        if (dispatchers.isEmpty()) break;
+                        else dispatchers.forEach(Dispatcher::sendBatchedRequests);
+                    }
+                } catch (InterruptedException e) {
+                    LOG.error(e.getMessage(), e);
+                } finally {
+                    isRunning.set(false);
+                }
+                if (!dispatchers.isEmpty()) mayStartRunning();
+            }
+        }
+
+        public void clear() {
+            dispatchers.clear();
+        }
+
+        public Dispatcher dispatcher(StreamObserver<TransactionProto.Transaction.Reqs> requestObserver) {
+            Dispatcher dispatcher = new Dispatcher(requestObserver);
+            dispatchers.add(dispatcher);
+            return dispatcher;
+        }
+
+        public class Dispatcher implements AutoCloseable {
+
+            private final StreamObserver<TransactionProto.Transaction.Reqs> requestObserver;
+            private final ConcurrentLinkedQueue<TransactionProto.Transaction.Req> requestQueue;
+
+            private Dispatcher(StreamObserver<TransactionProto.Transaction.Reqs> requestObserver) {
+                this.requestObserver = requestObserver;
+                requestQueue = new ConcurrentLinkedQueue<>();
+            }
+
+            private synchronized void sendBatchedRequests() {
+                if (requestQueue.isEmpty() || !isOpen) return;
+                TransactionProto.Transaction.Req request;
+                ArrayList<TransactionProto.Transaction.Req> requests = new ArrayList<>(requestQueue.size() * 2);
```

Comment:
Why `* 2`? Isn't `requestQueue.size()` itself enough capacity for this list?

Reply:
`requestQueue` can still be populated through `Dispatcher.dispatch(...)` after this ArrayList has been created. :) 

So, even multiplying it by 2 does not guarantee the capacity fits, but it can reduce the probability the array gets resized.